src/Microsoft.AspNetCore.Hosting/Internal/StartupLoader.cs

```diff
@@ -23,44 +60,170 @@ public static StartupMethods LoadMethods(IServiceProvider hostingServiceProvider
                 instance = ActivatorUtilities.GetServiceOrCreateInstance(hostingServiceProvider, startupType);
             }
 
-            var configureServicesCallback = servicesMethod.Build(instance);
-            var configureContainerCallback = configureContainerMethod.Build(instance);
+            // The type of the TContainerBuilder. If there is no ConfigureContainer method we can just use object as it's not
+            // going to be used for anything.
+            var type = configureContainerMethod.MethodInfo != null ? configureContainerMethod.GetContainerType() : typeof(object);
+
+            var builder = (ConfigureServicesDelegateBuilder) Activator.CreateInstance(
+                typeof(ConfigureServicesDelegateBuilder<>).MakeGenericType(type),
+                hostingServiceProvider,
+                servicesMethod,
+                configureContainerMethod,
+                instance);
+
+            return new StartupMethods(instance, configureMethod.Build(instance), builder.Build());
+        }
+
+        private abstract class ConfigureServicesDelegateBuilder
+        {
+            public abstract Func<IServiceCollection, IServiceProvider> Build();
+        }
+
+        private class ConfigureServicesDelegateBuilder<TContainerBuilder> : ConfigureServicesDelegateBuilder
+        {
+            public ConfigureServicesDelegateBuilder(
+                IServiceProvider hostingServiceProvider,
+                ConfigureServicesBuilder configureServicesBuilder,
+                ConfigureContainerBuilder configureContainerBuilder,
+                object instance)
+            {
+                HostingServiceProvider = hostingServiceProvider;
+                ConfigureServicesBuilder = configureServicesBuilder;
+                ConfigureContainerBuilder = configureContainerBuilder;
+                Instance = instance;
+            }
+
+            public IServiceProvider HostingServiceProvider { get; }
+            public ConfigureServicesBuilder ConfigureServicesBuilder { get; }
+            public ConfigureContainerBuilder ConfigureContainerBuilder { get; }
+            public object Instance { get; }
 
-            Func<IServiceCollection, IServiceProvider> configureServices = services =>
+            public override Func<IServiceCollection, IServiceProvider> Build()
```

Comment:
I'm closing over data. Local functions are just a nicer signature for Func<...,...> lambda = ....